File: samples/dotnet-service/src/Pricing/Order.cs
namespace Pricing;

/// <summary>Customer tier.</summary>
public enum Tier
{
    /// <summary>No tier discount.</summary>
    None,

    /// <summary>5% off.</summary>
    Silver,

    /// <summary>10% off.</summary>
    Gold,
}

/// <summary>Immutable order. Money is <see cref="decimal"/>, never double.</summary>
public sealed record Order
{
    /// <summary>Creates an order and validates it.</summary>
    public Order(decimal subtotal, Tier tier, int items)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(subtotal);
        Subtotal = subtotal;
        Tier = tier;
        Items = items;
    }

    /// <summary>Order subtotal before discounts.</summary>
    public decimal Subtotal { get; }

    /// <summary>Customer tier.</summary>
    public Tier Tier { get; }

    /// <summary>Number of items.</summary>
    public int Items { get; }
}

File: samples/java-gradle-service/src/main/java/com/example/pricing/Order.java
package com.example.pricing;

import java.math.BigDecimal;
import java.util.Objects;

/** Immutable order. Money is BigDecimal, never double. */
public record Order(BigDecimal subtotal, Tier tier, int items) {

    public Order {
        Objects.requireNonNull(subtotal, "subtotal");
        Objects.requireNonNull(tier, "tier");
        if (subtotal.signum() < 0) {
            throw new IllegalArgumentException("subtotal cannot be negative");
        }
    }
}
